File: java/src/test/java/com/dddeurope/niuz/fixtures/InMemoryEventBus.java
package com.dddeurope.niuz.fixtures;

import com.dddeurope.niuz.events.EventPublisher;
import com.dddeurope.niuz.events.Topic;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class InMemoryEventBus implements EventPublisher, Topic {
    private final List<Subscription> subscriptions = new ArrayList<>();

    @Override
    public <T> void subscribe(Class<T> eventType, Consumer<T> handler) {
        subscriptions.add(new Subscription<>(eventType, handler));
    }

    @Override
    public void publish(Object event) {
        for (Subscription s : subscriptions) s.handle(event);
    }

    private static class Subscription<T> {
        private final Class<T> eventType;
        private final Consumer<T> handler;

        public Subscription(Class<T> eventType, Consumer<T> handler) {
            this.eventType = eventType;
            this.handler = handler;
        }

        public void handle(Object event) {
            if (eventType.equals(event.getClass())) {
                handler.accept((T) event);
            }
        }
    }}

File: dotnet/niuz.application.test/fixtures/InMemoryEventBus.cs
using System;
using System.Collections.Generic;
using niuz.application.events;

namespace niuz.application.fixtures
{
    public class InMemoryEventBus : ITopic, IEventPublisher
    {
        private readonly List<IEventSubscription> subscriptions = new List<IEventSubscription>();

        public void Subscribe<T>(Action<T> handler) where T : class
        {
            subscriptions.Add(new EventSubscription<T>(handler));
        }

        public void Publish(object @event)
        {
            foreach (var subscription in subscriptions) subscription.Handle(@event);
        }

        private interface IEventSubscription
        {
            void Handle(object @event);
        }

        private class EventSubscription<T> : IEventSubscription where T : class
        {
            private readonly Action<T> handler;

            public EventSubscription(Action<T> handler)
            {
                this.handler = handler;
            }

            public void Handle(object @event)
            {
                if (@event.GetType() == typeof(T)) handler(@event as T);
            }
        }
    }
}